### ISSUE TITLE
No automatic updates if GIT_TAG points to a branch 

### ISSUE THREAD

**Issue by DmitryYurov** [cpm-cmake/CPM.cmake#263]
I am trying to use cpm for automating dependencies of my projects.
Unfortunately if I use cpm like
```
CPMAddPackage(
  NAME name_of_repo
  GIT_REPOSITORY ssh://git@server/project.git
  GIT_TAG master
  EXCLUDE_FROM_ALL YES
)
```
the fetched resource does not track the master branch - that is, on update of my master branch I still get older commit in the local repository. On the other hand, if I manually remove the fetched data, new download looks at the proper commit.

If that matters, I set `CPM_SOURCE_CACHE` environment variable.
Is it a bug or expected behaviour? I would expect that resources are downloaded again if the branch state in remote repository changes.

**Comment by TheLartians** [comment]
This is expected behaviour, as CPM.cmake assumes that the `TAG` argument point to an unchangeable git tag or commit hash. We encourage this, as otherwise builds would become unstable, e.g. they would change every time the branch is modified. Also this allows us to efficiently cache sources and have a predictable offline and online behaviour. 

For parallel development of dependency / dependant projects, I would either manually update the tag to the most recent commit or use a [local package override](https://github.com/cpm-cmake/CPM.cmake#local-package-override) to pull the source from a local directory that is kept up to date with the master branch.

**Comment by DmitryYurov** [comment]
In this case the functionality related to branch-named tag seems to be internally contradictory - on the one hand if I reference the tip of the branch, I would like the reference to stay on this tip - but it is fixed on the last commit at the moment of the first dependency download. On the other hand, this tag does not provide the stability, since the user could for example setup a clean build on the server and in such a way update the reference all the time. The behavior of fixed dependency in case of a version tag looks good, but with branch-named tag it is quite unexpected.

**Comment by TheLartians** [comment]
You're right, the README is actually wrong in this case as it was written before we implemented caching and offline support. We actively discourage the use of branch names, as it would prevent us from supporting offline builds and reproducibility due to the reasons mentioned above. Perhaps we could add an alternative option `BRANCH_NAME` that would allow the old functionality of auto-updating branches.

**Comment by N3m3si5** [comment]
> You're right, the README is actually wrong in this case as it was written before we implemented caching and offline support. We actively discourage the use of branch names, as it would prevent us from supporting offline builds and reproducibility due to the reasons mentioned above. Perhaps we could add an alternative option `BRANCH_NAME` that would allow the old functionality of auto-updating branches.

googletest recommends to use latest master: see [Live at Head](https://github.com/google/googletest/tree/master). Therefore I personally think this is a handy feature in some circumstances. 👍

**Comment by anb0s** [comment]
I've tried CPM and it works very well! Thank You for sharing! (imho: it could be official part of the CMake distribution ;)

The only issue is to support work with multiple dependencies as local repositories "in work" and referencing therefore latest branch commits for "integrators". Here we want to have automatic updates from latest branches and set many tags as workaround does not feel right ;) During the merge to mainline we verify that only tags or commits are used in such "read only" repositories (dependencies) for all other developers.

Do you have plans to introduce BRANCH_NAME to support this behaviour?

**Comment by TheLartians** [comment]
Thanks for the kind feedback, glad it's useful for you! :) 

> Do you have plans to introduce BRANCH_NAME to support this behaviour?

Currently no plans, though what speaks against just using plain old [`FetchContent`](https://cmake.org/cmake/help/latest/module/FetchContent.html) in this case? I believe it should match the expected behaviour for branches - including loosing most advantages of CPM.cmake anyways (no caching, no offline builds, slow to configure etc).

**Comment by anb0s** [comment]
> Thanks for the kind feedback, glad it's useful for you! :)
> 
> > Do you have plans to introduce BRANCH_NAME to support this behaviour?
> 
> Currently no plans, though what speaks against just using plain old [`FetchContent`](https://cmake.org/cmake/help/latest/module/FetchContent.html) in this case? I believe it should match the expected behaviour for branches - including loosing most advantages of CPM.cmake anyways (no caching, no offline builds, slow to configure etc).

Yes, sure, that's how we handle this - we are using submodules or FetchContent. I just tried to have CPM as the dependency manager in all places independent how the dependency is consumed (remote location etc.) and checked out like local path (sub-dir) in source tree or ouside source tree e.g. in the build folder. At the end i want to have simple clone without dependencies (no submodules etc.) and CMake configure should handle this via CPM :)

**Comment by DNKpp** [comment]
If I understand you correctly, your main reason against the auto updating is, that this doesn't perform well in an offline environment. That's completely understandable but I wonder if a ``step-in-the-middle`` could help. If one target a branch, why not simply *try* making a connection and see if the origin has been updated. If no connection is possible, than simply use the local version. That's what I expect if I use a cache.

**Comment by OlivierLDff** [comment]
From experience I wouldn't recommand using git branch, because then it's hard to track in the life of the project which dependency is expected where.
Also, if a branch name change, or is deleted, then you are just left with an error. All of my old commit won't compile with `googletest` because they change `master` to `main`. So each time I go to an old commit I need to fix that. I would have used git tag back then I would be fine.

Lesson learned: 
- Never use public github repository, always fork it to be sure what you have in there (and then you can maintain branch if you want). That will also help avoid malicious commit, deleted branch, repo deleted by user, etc...)
- Use git tag when you are in control of the repository (ie you forked it). So that you can clearly see in your commit history the dependency that evolve.
- Or use commit sha, if you are develloping the dependency library at the same time.

Expecting auto update will only lead to commit not compiling in the future, and making tools like `git-bisect` ineffective.

**Comment by TheLavaBlock** [comment]
Hey, in my library [liblava](https://git.io/liblava) I integrate external modules rolling release-like, so I solved it with a simple python script `update.py` which updates all GIT_TAG when the opportunity arises. If you are interested, please take a look. It would be great if there was such a mode in CPM.cmake and it all worked automatically.

**Comment by TheLartians** [comment]
@TheLavaBlock that's a clever solution, thanks for sharing! I completely agree that this would be a great feature to have built into CPM.cmake itself, but the implementation would be a bit tricky due to edge cases and the imperative nature of CMake itself.

Note that we have a related discussion on this feature [here](https://github.com/cpm-cmake/CPM.cmake/discussions/306).

**Comment by derekcyruschow-catapult** [comment]
> Hey, in my library [liblava](https://git.io/liblava) I integrate external modules rolling release-like, so I solved it with a simple python script `update.py` which updates all GIT_TAG when the opportunity arises. If you are interested, please take a look. It would be great if there was such a mode in CPM.cmake and it all worked automatically.

https://joht.github.io/johtizen/automation/2022/08/03/keep-your-cpp-dependencies-up-to-date.html#how-to-use-renovate-with-cpmcmake sounds like a similar approach using https://github.com/renovatebot/renovate

**Comment by Napolitain** [comment]
> From experience I wouldn't recommand using git branch, because then it's hard to track in the life of the project which dependency is expected where. Also, if a branch name change, or is deleted, then you are just left with an error. All of my old commit won't compile with `googletest` because they change `master` to `main`. So each time I go to an old commit I need to fix that. I would have used git tag back then I would be fine.
> 
> Lesson learned:
> 
> * Never use public github repository, always fork it to be sure what you have in there (and then you can maintain branch if you want). That will also help avoid malicious commit, deleted branch, repo deleted by user, etc...)
> * Use git tag when you are in control of the repository (ie you forked it). So that you can clearly see in your commit history the dependency that evolve.
> * Or use commit sha, if you are develloping the dependency library at the same time.
> 
> Expecting auto update will only lead to commit not compiling in the future, and making tools like `git-bisect` ineffective.
> 
> Edit: auto update should be handled by a tool like dependabot

That is wrong.
Staying at HEAD means you know WHEN it breaks rather than lock your behavior to older version due to laziness and then taking 2 months to actually understand which versions really breaks your code, and why. 
Abseil at google already stated many reasons to prefer live at HEAD and allowing all options is better than allowing an outdated way of thinking on others.
Btw I am curious which version of zlib you currently use since you maintain your own fork very well.
